### ISSUE TITLE
Have Locals code add locals below header block and below lParameter statements

### ISSUE THREAD

**Issue by avianmanagement** [VFPX/Thor#104]
I've also started to use Thor recently to add m. in all my code.

When I run it it also picks up if I have missed any local variables and adds them, very nice.

It adds them right at the top of my code.

I have a commented header block at the top of all my code.

This looks like this:

#If .F.
	Text
********************************
*   HELP BUILDER COMMENT BLOCK *
********************************
*:Help Documentation
*** Copyright Avian Management Services 2005-- {^2005-11-25 17:15:22}
*:Name:
aim_get_accession_code
*:Purpose:
Returns the unique accession number for a bird
*:Keywords:

*:Parameters:

*:Returns:

*:Remarks:

*:Example:

*:EndHelp
	Endtext
#Endif
lParameters tcpk_dBird
Local lcBirdPK

It puts the new locals right at the top, above the #If .F..  

I have to move it down below the comments and next to the exiting local declarations.

Is there a way to get Thor to place it with any other locals already in the code, such as in the line in the example above Local lcBirdPK

If there are no locals already in the code could it add them after any headed block and also very importantly if one has an lParameters at the start of the code it needs to be below that.

At the moment when ever I run the beautfy in thor that add mdot and locals I have to check VERY carefully that there is no lParameter line in the code  or I will get an error when the code runs as there will be a local line above the lParametsr statement.

**Comment by Jimrnelson** [comment]
David --

When working on "Create Locals", years ago, I was amazed to learn of the variety of ways that LOCAL statements are created, how important the format is (to at least some users) and also how important the placement in the code is.

Thus it is that I realize that your request creates quite a conundrum, as I have to be extremely careful not to offend other users who are particular about where the LOCAL statements go.

The one rule I am sure of (and is not now implemented) is that the LOCAL statement must come **after** any PARAMETER or LPARAMETER statement.  Putting it before obviously creates a bug.  (I am wondering if there are any who avoid using "Create Locals" specifically because it would create this bug.)

Thus, I can see that I could handle your request by implementing this rule (it's a bug fix , so nobody can argue with it) which would help for any procedures of yours that have a PARAMETER or LPARAMETER statement.  In fact, since it's a bug, it should be done whether you requested it or not.

This doesn't help your case when you don't have a PARAMETER or LPARAMETER statement.  I will have to consider further how to implement that without changing behavior for anybody else.

**Comment by avianmanagement** [comment]
Hi Jim,

The main issue is that it needs to come after the Parameter or lParamater statement as it breaks code.
I have had it twice now where I have missed it and it caused errors of no parameter statement found.

The after comment block would be a nice to have, but not a bug as such.

The add it to other local statements rather than at the very top again is a nice to have, but could maybe be implemented via a setting to the user can mark it to use or not, as it seems to be a very logical thing to do.

Using  BeautifyX with Mdot and locals turned on after witting a load of code is brilliant and such a time saver.

I've just spent two days going through all the code in my main application and running this on code block.  It picked up at least five instances where I had missed a local declaration.

**Comment by lscheffler** [comment]
It works for ages with me. Something odd / special  with your LPARAMETERS block? Like abbreviations, comments, AS, empty lines or so? All my LPARAMETERS follow the form
```
LPARAMETERS;
 txPara1,;
 txPara2,;
 txPara3
```

The settings where to create LOCALS are part of *Create Locals* not * BeautifyX*. If you uncheck "Move/Merge .. to top*, they will be created next to first occurrence.

**Comment by avianmanagement** [comment]
Thank you

I see now that if the first line in my PRG is the parameter statement it works.

My problem is all of my Prgs, and other code, has a header block for documentation of the code then the parameter statement, so it goes to the very top of the PRG every time it adds any.

For now I'll add my parameter statements at the top, but if it were possible for it to add these below head blocks it would be really nice.

**Comment by avianmanagement** [comment]
Testing this some more it has some very odd behaviour

Sample code
![image](https://github.com/VFPX/Thor/assets/5089118/0641e095-e35a-42ff-af35-ab5345aa5335)

Run beautify which calls the locals code gives

![image](https://github.com/VFPX/Thor/assets/5089118/b25baadb-d456-477c-86d4-d6244bc3dd04)

Add new local variable below my lctest1 variable and run it again and i get

![image](https://github.com/VFPX/Thor/assets/5089118/03668087-678c-401c-be07-0b16823ef19a)

It has removed all my comment block except the last #ENDIF

**Comment by lscheffler** [comment]
@Jimrnelson This looks a bit like a bug. (The one with the comment at top) Me or you? I have done some work on this. While I hate it, I will find it.

**Comment by lscheffler** [comment]
@avianmanagement This looks related, but is a different problem. Please be kind, start a new issue and edit the message above to point to new issue.

**Comment by Jimrnelson** [comment]
@lscheffler 

You can take this if you want; even though PEMEditor was my first tool, I have not touched it in ages.

Unfortunately, there are a lot of tools embedded in PEMEditor that really have nothing to do with the PEMEditor form.  This is due to the evolution of PEMEditor, eventually spawning Thor.

I see that there are at least two problems here:

[1]  There needs to be an initial pass through the code to look for a Parameter/LParameter statement.  If any, the LOCAL statements created must follow.  This is new code, as currently there is only one pass.  The existing code assumes that the Parameter/LParameter statement, if any, is the first statement.

[2]  It looks like the #IF / #Endif block got swallowed, which must be some new bug.  The code looks for #IF but really should be looking for #IF .F.

**Comment by lscheffler** [comment]
@Jimrnelson Yes, as I wrote avianmanagement, this should be two issues. Let's see if he is splitting it, or I will do.

It just depends on your workload / fun with it. As far as I see, some changes to it where made by me, so I can do it again. Just let me know.

And yes, this is a mine field of code. But I've see much worst then this. :)

**Comment by Jimrnelson** [comment]
Tis indeed a minefield, which happens when there are many contributors.

I'd be happy to work on it if it were not part of PEMEditor and could be worked on separately.  That is true for quite a few tools which originated as part of PEMEditor before we created Thor,

**Comment by lscheffler** [comment]
So. I do. No problem.

**Comment by Jimrnelson** [comment]
#DEFINEMaybe   #If .F.    structureSent from my iPhoneOn May 26, 2023, at 10:05 AM, Lutz Scheffler ***@***.***> wrote:﻿
Do we agree the following for placing LOCALs on top:
LOCALs will be written before the first line that is not

a comment
empty
#INCLUDE
LPARAMETERS / PARAMETERS (block),

but, if a LPARAMETERS / PARAMETERS block is found, immediately after this?

**Comment by avianmanagement** [comment]
It will always need to go after LPARAMETERS / PARAMETERS

The #If .F.  is normally the first line of my comments section

![image](https://github.com/VFPX/Thor/assets/5089118/9beedd1a-4d32-4e5a-b29d-ce2482d4232b)


If the first line is #If .F. and then there is a comment block that ends with after which there is no LPARAMETERS / PARAMETERS can it go after the #If Endif of the comment block

**Comment by lscheffler** [comment]
If there is a LPARA (..) then below LPARA, if not above the #IF, by the rules of the message before.  Since #IF #ELSE #ENDIF could hold code running, there is no way to determine if it is active.

Your use of #IF for a comment is syntactically valid, but could not be generalized.

**Comment by avianmanagement** [comment]
OK thank you

**Comment by Jimrnelson** [comment]
Except you CAN check for #if .F.Which is a common enough construct.  Sent from my iPhoneOn May 26, 2023, at 10:42 AM, Lutz Scheffler ***@***.***> wrote:﻿
If there is a LPARA (..) then below LPARA, if not above the #IF, by the rules of the message before.  Since #IF #ELSE #ENDIF could hold code running, there is no way to determine if it is active.
Your use of #IF for a comment is syntactically valid, but could not be generalized.

**Comment by Jimrnelson** [comment]
No, that’s not right either, because that will change current behavior for others.I will elaborate in a bit.  Sent from my iPhoneOn May 26, 2023, at 10:42 AM, Lutz Scheffler ***@***.***> wrote:﻿
If there is a LPARA (..) then below LPARA, if not above the #IF, by the rules of the message before.  Since #IF #ELSE #ENDIF could hold code running, there is no way to determine if it is active.
Your use of #IF for a comment is syntactically valid, but could not be generalized.

**Comment by lscheffler** [comment]
> No,  that’s not right either, because that will change current behavior for others.I will elaborate in a bit.

This **is** current behaviour, infact in current code `#IF` forces explicit to insert the LOCALs before.

See `MODIFY CLASS beautify OF peme_beautify.vcx METHOD DoLocals`

Basically this is why the LOCALS are inserted there now. It's the use of #IF for a comment that fails, a comment like * will run perfectly well and the LOCALs will be placed after LPARA. So, my proposed change will keep behaviour if no LPARA exists, but including #INCLUDE statements to the stuff ignored.

I call this no longer a bug, but an improvement, *Create LOCALs* works perfectly in the limits one would expect, since it's to special to fiddle with #IF statements. 

BTW, there is no need for an * initial pass*. I just need to figure out where to place the string and use a STUFF after the loop is done. So it's either the first possible location, or the location after the parameters block. I love STUFF. :)

**Comment by Jimrnelson** [comment]
As I said, I will elaborate later, when I have time to do so. Sent from my iPhoneOn May 26, 2023, at 11:55 AM, Lutz Scheffler ***@***.***> wrote:﻿

No,  that’s not right either, because that will change current behavior for others.I will elaborate in a bit.

This is current behaviour, infact in current code #IF forces explicit to insert the LOCALs before.
See MODIFY CLASS beautify OF peme_beautify.vcx METHOD DoLocals
Basically this is why the LOCALS are inserted there now. It's the use of #IF for a comment that fails, a comment like * will run perfectly well and the LOCALs will be placed after LPARA. So, my proposed change will keep behaviour if no LPARA exists, but including #INCLUDE statements to the stuff ignored.
I call this no longer a bug, but an improvement, Create LOCALs works perfectly in the limits one would expect, since it's to special to fiddle with #IF statements.
BTW, there is no need for an * initial pass*. I just need to figure out where to place the string and use a STUFF after the loop is done. So it's either the first possible location, or the location after the parameters block. I love STUFF. :)

**Comment by lscheffler** [comment]
@avianmanagement 
> OK thank you If I have
> # IF .F.
> 
> ** code block here #Endif Local lcTest
> 
> Then would other locals that are generated by Thor be added to the existing line for local lcTest, or would they goto the top of the prg?

Right now what comes first, here #IF. 

The idea for "On Top" is - on top. If one would search for an existing LOCAL, it adds a lot of complexity to keep it "on top". One can not trust that the LOCAL is on top of any other code.

Also, if **I** use `#IF .F.` I like my locals on top of it, because it's usually a switch for debugs.

It might be cheaper to add an otherwise useless `LPARA tvNothing` to code without parameters to solve your problem.  As soon as the LOCALS are forced below the parameters block.

**Comment by avianmanagement** [comment]
Yes I have considered the LPARA tvNothing, which would solve the issue on this.

I can add that to the intellisence code that adds the code comment block.
That would be a good workaround if you can implement the locals being placed after the first lpara statment in the code, which in my case would be after the comment block

**Comment by Jimrnelson** [comment]
Can we push the discussion for #if to a new issue?Sent from my iPhoneOn May 26, 2023, at 1:14 PM, Lutz Scheffler ***@***.***> wrote:﻿
@avianmanagement

OK thank you If I have
IF .F.
** code block here #Endif Local lcTest
Then would other locals that are generated by Thor be added to the existing line for local lcTest, or would they goto the top of the prg?

Right now what comes first, here #IF.
The idea for "On Top" is - on top. If one would search for an existing LOCAL, it adds a lot of complexity to keep it "on top". One can not trust that the LOCAL is on top of any other code.
Also, if I use #IF .F. I like my locals on top of it, because it's usually a switch for debugs.
It might be cheaper to add an otherwise useless LPARA tvNothing to code without parameters to solve your problem.  As soon as the LOCALS are forced below the parameters block.

**Comment by lscheffler** [comment]
@Jimrnelson see #107

**Comment by Jimrnelson** [comment]
I think this list is fine, with a couple of additions (which you might already have recognized):

> Do we agree the following for placing LOCALs on top:
> 
> LOCALs will be written before the first line that is not
> 
> * a comment
> * empty
> * #INCLUDE
> * #DEFINE
> * PUBLIC
> * PRIVATE
> * LPARAMETERS / PARAMETERS (block),

But I disagree with this:

> but, if a LPARAMETERS / PARAMETERS block is found, immediately after this?

I think you just follow the first rule --- I would like all of those things to precede my LOCAL statements regardless of the position of my PARAMETERS statements.

I believe the current code does NOT put LOCSALs immediately after LPARA/PARA, and we don't want to change this.

I am beginning to wonder what, if anything needs to be changed ... is it just #IF .F.   ??

**Comment by lscheffler** [comment]
Given the option *Locals to top* is set, the version life on github will do exactly as I wrote in https://github.com/VFPX/Thor/issues/104#issuecomment-1564446775, except it will put the local in front of the first #if if there is a parameters statement or not. 

IOW. Now the LOCALs are be placed in following order, whatever comes first
1. in front of the first #IF
2. if a parameters block is found, directly after
3. in front of any line, that is not starting with \*. \# or && or is empty.

All my change is, that I look up the parameters block and give it precedence, so then it will look like
2. if a parameters block is found, directly after, as long as the block exists
1. in front of the first #IF or
3. in front of any line, that is not starting with \*. \# or && or is empty, whatever comes first

**Comment by Jimrnelson** [comment]
No, what you have written here is not good enough:

> 1. if a parameters block is found, directly after, as long as the block exists
> 2. in front of the first #IF or
> 3. in front of any line, that is not starting with *, #, NOTE or && or is empty, whatever comes first

You cannot assume that once you see PARAMETERS you can simply plunk in LOCALS immediately thereafter.  That is certainly not how it works now.

(Most of my code has PARAMETERS as the first line, and my LOCAL statements appear after comments, #DEFINEs, etc.)

You should put LOCALs after the greater of:
* the line immediately before all the special non-executable code we have been mentioning (comments, empty, and so on)
* the line immediately after a PARAMETER statement

This will handle the cases currently handled where the PARAMETER statement comes early.

**Comment by lscheffler** [comment]
You are right

**Comment by lscheffler** [comment]
Now it is like 

1. General
    - in front of the first #IF
    - in front of any line, that is not starting with *, NOTE, #, && or is empty.
2. if a parameters block is found, *General* applies, but after the parameters block

**Comment by Jimrnelson** [comment]
I think you should add PUBLIC and PRIVATE to your list, along with NOTE.  I find it interesting that you say “parameters block”.  By that do you mean “parameters statement” ?Sent from my iPadOn May 27, 2023, at 4:49 AM, Lutz Scheffler ***@***.***> wrote:﻿
Now it is like

General

in front of the first #IF
in front of any line, that is not starting with *, NOTE, #, && or is empty.


if a parameters block is found, General applies, but after the parameters block

**Comment by lscheffler** [comment]
Parameters statement is the right expression. Only I think in blocks belonging together rather then statements. Because of the *inner statements* of a SELECT SQL, where I tend to get confused. Or Text .. EndText blocks, Procedures

If LOCAL before PUBLIC or the other way around is minor, and if *Create Locals* identifies something as global (i.e., not local) it is placed after the LOCALs as a comment, so I would keep it as it is.

**Comment by Jimrnelson** [comment]
> so I would keep it as it is.

Yes, it's very important to keep things as they are.  If at all possible, existing behavior should not change.

With that in mind, I notice the following behaviors using my current version of BeautifyX:

[1]  PUBLIC and PRIVATE should be in your list.  If not, the LOCAL statement here would have moved to after PRIVATE

![image](https://github.com/VFPX/Thor/assets/81179335/efbb8f3a-5f58-43ef-9d67-dd398fb96a0f)

[2]  If there is already a LOCAL statement, but before where your new code would have placed it, it remains in the same place.
Thus, moving the LOCAL statement up in the above code will leave it in place.

![image](https://github.com/VFPX/Thor/assets/81179335/b4168cb1-abd5-409e-b7dd-2875c2eab10e)

In reviewing all of what has gone in this thread, I believe that the only change we should see has to do with the mis-handlng of #IF.  Even the discussion about PARAMs is a red herring, as the only reason that is part of the discussion is because of @avianmanagement 's use of #IF .F.

Instead of re-writing all the code, this might have been as simple as adding an option to treat #IF .F. structures two different ways.

**Comment by lscheffler** [comment]
Possibly we argue about nothing. I just make sure that if a parameter block is found, LOCALs will be placed behind, in the manner it was before. If that was behind PUBLIC etc, it will be there, if not, not. And I guess I missed to list this part. The important point was *not direct after parameters* from above, this is done.

I will do nothing dealing with `#IF .F.` here, this is just `#IF` and will be handled as such. Same idea as above, change nothing, just no LOCALs above parameters. See #107 about it.

---
The basic idea is, it searches for places to insert LOCALs the old way, but not insert, just noting the first possible position. Reset this if params is found, and search again. STUFF after the whole blocked is searched. Since the only possible thing (syntactical) above params forcing an insertion of LOCALs is `#IF`; the location will be set, but then removed on params.   
If no params, position is perfect.

**Comment by Jimrnelson** [comment]
OK, I think I understand what you are doing now.

I was misled when you were wrote this, which made it appear that you were re-writing existing code:

> Now it is like
> 
> 1. General
>    
>    * in front of the first #IF
>    * in front of any line, that is not starting with *, NOTE, #, && or is empty.
> 2. if a parameters block is found, _General_ applies, but after the parameters block

I would hope you would be doing this:

_If there is a PARAM statement, start processing there; else start processing, as before, at the top of the procedure._

With the addedum that you are fixing the bug with the handling of #IF structures.

**Comment by lscheffler** [comment]
It's the like, start processing at top, and reset if there is a parameters statement. So it's just a straight run.

**Comment by avianmanagement** [comment]
Not sure if it this or 106 update but M dot does not seem to be working anymore

**Comment by lscheffler** [comment]
@avianmanagement please;
1. New problem, new issue
2. One problem, one issue (if possible)
3. Please be more specific

Until that, I see no problem, works as expected.

**Comment by avianmanagement** [comment]
I had a setting in the options  to only use where needed so not all m dots were being added, so not issue

**Comment by lscheffler** [comment]
@avianmanagement Thanks for the participation. Nice that it works again. I saw there where some changes to not call mDot in the middle of a string, possibly it was related to that fix.

If you run into a problem with mDots, LOCALS, or BeautifyX please open an issue at https://github.com/VFPX/PEMEditor/issues. I know, this is a problem to figure out, but it helps to trace the problems.   


@DougHennig , @Jimrnelson may we add with a list of tools that belong to PEMEditor to the Thor bugs template, so reporters will be pointed to this? I look up if one might the templates site to do it there. Thank you.

**Comment by Jimrnelson** [comment]
@lscheffler 
 
> @DougHennig , @Jimrnelson may we add with a list of tools that belong to PEMEditor to the Thor bugs template, so reporters will be pointed to this? I look up if one might the templates site to do it there. Thank you.

I would prefer not to point users to such a list.  I think we want to make it as easy as possible for them to report issues and we can sort out later which VFPX project is being referred.

I suggest instead that we ask users to report issues/bugs to the VFPX project, if they know it, otherwise just to Thor.  And, at the same time, asking that they provide as much information as possible about their issue/bug.

**Comment by lscheffler** [comment]
Since I do only maintain projects I'm working on I rather dislike this idea. Neither do I like to be spamed with any project, nor do I like to maintain issues outside their repository, all function provided in issues, commits, merges will be lost. This issue I found by accident.

If you like to do that, you might check what the issue-YAML-solution can do, possibly it can scatter from a project to it's repositories, I have no clue.

**Comment by Jimrnelson** [comment]
@lscheffler @DougHennig 

My primary concern is to not provide any obstacles that would get in the way of users of submitting issues / bug reports.

I think it's fine to ask them to identify the relevant VFPX project, if possible, which is why I wrote:

> I suggest instead that we ask users to report issues/bugs to the VFPX project, **if they know it**, otherwise just to Thor. 

But I would not stress this.  There are tools where it is not at all easy (even for me) to determine the relevant project.

Beyond this, it is clear to me over the years that users think along the lines of "It's all Thor".  

Today we have a fine example of this.  Tom Knauf ,who clearly knew he was using IntellisenseX (it's in his subject line) added the issue to Thor.  I would say this is the norm, not the exception.

If you don't receive notifications when issues are raised in Thor, we will be happy to notify you of any issues in projects you support.

**Comment by lscheffler** [comment]
google on [github issue template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository) or [github issue yaml](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/syntax-for-issue-forms) shows you the more complex YAML way to create a kind of issue template or whatever,  more complex then an MarkDown template.
